### PR TITLE
[Release] Add Nightly Release workflow implementation 

### PR DESCRIPTION
**Pull request by alexander-smolyakov**
**Description**:
This pull request adds implementation for the Nightly Release workflow.

**Workflow description**:
Workflow triggers:
- On schedule;
- Manual launch;

The `aggregate-release-data` stage: 
- Resolve the latest commit from the `main` branch, and set the `commit_sha` output;
- Check version number in sources, parse and verify it;
- Generate a new version based on the version in sources, and set `version_number` output;

The `release-github-pypi` stage:
- Pass data from the `aggregate-release-data` stage to release workflow and call it;

**Changelog**:
_nightly-release.yml_:
- Added initial implementation;

_release.yml_:
- Added `workflow_call` trigger;

_.bumpversion.cfg_:
- Updated `parse` and `serialize` properties of config to respect the development segment;

**Versioning strategy for nightly releases**:
According to [PEP 440](https://peps.python.org/pep-0440/), we cannot set the release version as nightly. However, we can add the **development** and **local** segments to the current version. This will allow differentiating nightly releases from test and prod releases. So that version of the nightly build version number should look like `v1.4.0a1.dev01042023+nightly`, and the normalized version will look the following: `1.4.0a1.dev1042023`. [Version normalization](https://peps.python.org/pep-0440/#normalization) happens during publishing to the package index. During the version normalization, we will lose the local segment of the version but will preserve the development segment.

_Example_:
| Version in sources | Generation strategy              | Nightly release version       |
| ------------------ | -------------------------------- | ----------------------------- |
| `1.4.0a1`          | `{ version }.dev{m/d/Y}+nightly` | `1.4.0a1.dev01042023+nightly` |

**Work example**:
- https://github.com/dbt-labs/dbt-core-release-test/actions/runs/3840385988

### CODE REVIEW

**Comment by alexander-smolyakov** [comment]
## Example of work
**Workflow run**: https://github.com/dbt-labs/dbt-core-release-test/actions/runs/3840385988

**Workflow log**:
- The `prep-release/test-run/1.4.0a1.dev01042023+nightly_3840385988` branch created
- Changelog file created successfully
- Version successfully bumped in codebase to `1.4.0a1.dev01042023+nightly`
- S3 build artifact path: `s3://core-team-artifacts/dbt-labs/dbt-core-release-test/artifacts_testing/1.4.0a1.dev01042023+nightly/3d92787e32bb3351a754507caa6c7d01c7ded26a`
- Draft release: https://github.com/dbt-labs/dbt-core-release-test/releases/tag/untagged-119c0caba1b4aedf3f5a

**Published packages**: 
- https://test.pypi.org/project/dbt-tests-adapter/1.4.0a1.dev1042023/
- https://test.pypi.org/project/dbt-core/1.4.0a1.dev1042023/
- https://test.pypi.org/project/dbt-postgres/1.4.0a1.dev1042023/

File: .github/workflows/nightly-release.yml [hunk 1]
```diff
@@ -1,21 +1,92 @@
...
+  workflow_dispatch: # for manual triggering
+  schedule:
+    - cron: 0 2 * * *
 
 defaults:
   run:
     shell: bash
 
+env:
+  RELEASE_BRANCH: "users/alexander-smolyakov/implement-nightly-releases-workflow"
```
Comment (emmyoop) [inline .github/workflows/nightly-release.yml]: What do you see this being moving forward? `main`?

**Review by emmyoop — APPROVED** [review]

File: .github/workflows/nightly-release.yml (hunk 1 above)
Comment (alexander-smolyakov) [reply .github/workflows/nightly-release.yml]: This branch was used for testing. I will switch to the main branch before merging it.

Comment (alexander-smolyakov) [reply .github/workflows/nightly-release.yml]: Fixed in 6c65703a1f9c9a25e327a27ba0e8b0127ed1533a

**Pull request merged**